spec: Variant lower/upper bounds (#12658)
apache/iceberg: format/spec.md

5. The `content_offset` and `content_size_in_bytes` fields are used to reference a specific blob for direct access to a deletion vector. For deletion vectors, these values are required and must exactly match the `offset` and `length` stored in the Puffin footer for the deletion vector blob.
6. The following field ids are reserved on `data_file`: 141.

> For Variant, values in the `lower_bounds` and `upper_bounds` maps store the a serialized Variant object that contains lower and upper bounds for fields within the Variant. The object keys are normalized JSON path expressions that uniquely identify a Variant field. The object values are primitive Variant representations of the lower or upper bound for the field. Including bounds for any field is optional and the bounds must have the same Variant type.

Reviewer: ```suggestion
For Variant, values in the `lower_bounds` and `upper_bounds` maps store serialized Variant objects that contain lower or upper bounds respectively. The object keys for the bound-variants are normalized JSON path expressions that uniquely identify a field. The object values are primitive Variant representations of the lower or upper bound for that field. Including bounds for any field is optional and the bounds must have the same Variant type as the field being bounded.
```

Reviewer: Some grammar fixes as well as some wording changes. Semantically the same

> Bounds for a field must be accurate for all non-null values of the field in a data file. Bounds for values within arrays must be accurate all values in the array. Bounds must not be written to describe values with mixed Variant types (other than null). For example, a "measurement" field that contains int64 and null values may have bounds, but a string value such as "n/a" or "0" in any record would cause the bounds to be skipped.

Reviewer: ```suggestion
Bounds for a field must be accurate for all non-null values of the field in a data file. Bounds for values within arrays must be accurate all values in the array. Bounds must not be written to describe values with mixed Variant types (other than null). For example, a "measurement" field that contains int64 and null values may have bounds, but if the field also contained a  string value such as "n/a" or "0" then bounds may not be stored.
```

Reviewer: I want to be sure, but the last sentence should be a rule correct? You may not store bounds if multiple types exist for the field

Reviewer 2: That's correct. The problem with "0" is that it assumes engine behavior if we either cast it to 0 or if we don't. We want to match the engine's definition of when a row matches, which means being conservative and leaving casting an engine responsibility.

The Variant bounds objects are serialized by concatenating the [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md) of the metadata (containing the normalized field paths) and the bounds object.
Field paths follow the JSON path format to use normalized path, such as `$['location']['latitude']` or `$['user.name']`. The special path `$` represents bounds for the variant root, indicating that the variant data consists of uniform primitive types, such as strings.

Examples of valid field paths using normalized JSON path format are:

* `$` -- the Variant root value
* `$['user.name']` -- the field `"user.name"` in the root value that is a Variant object
* `$['location']['latitude']` -- the field `latitude` in a nested `location` object

> * `$['ids']` -- the `ids` array

Reviewer 2: When I read the thread you raised on the dev list, I liked that you used "tags" as the example. Maybe we should change some of these to match the examples in the variant shredding spec?

Author: Let me do that to match as much as possible.

5. The `content_offset` and `content_size_in_bytes` fields are used to reference a specific blob for direct access to a deletion vector. For deletion vectors, these values are required and must exactly match the `offset` and `length` stored in the Puffin footer for the deletion vector blob.
6. The following field ids are reserved on `data_file`: 141.

> For `variant` type, the `lower_bounds` and `upper_bounds` store the lower and upper bounds for all shredded subcolumns within a file. These bounds are represented as a Variant object, where each subcolumn path serves as a key and the corresponding bound value as the value. The object is then serialized into binary format (see [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md)).

Reviewer 2: This should not state "all shredded subcolumns". First, "all" is not a requirement and is actually misleading because the fields are only those for which we can guarantee the lower and upper bounds are accurate. I would also not use the somewhat confusing term "subcolumn". In the variant spec we refer to fields. You may also want to call out a special case where the root is not an object. In that case the lower or upper bound is tracked by the field name `"$"` indicating the root.

> For `variant` type, the `lower_bounds` and `upper_bounds` store the minimum and maximum values for all shredded subcolumns within a file. These bounds are represented as a Variant object, where each subcolumn path serves as a key and the corresponding bound value as the value. The object is then serialized into binary format (see [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md)).

Reviewer 3: Since these are bounds and not true minimum/maximum (correct me if I'm wrong if variant shredding is somehow different), should we change the wording to simply say "store the lower and upper bounds for all shredded sumbcolumns within a file"?

Author: It means the same as the other bounds and it may not be the actual min/max. Let me update that.

> For `variant` type, the `lower_bounds` and `upper_bounds` store the minimum and maximum values for all shredded subcolumns within a file. These bounds are represented as a Variant object, where each subcolumn path serves as a key and the corresponding bound value as the value. The object is then serialized into binary format (see [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md)).

Reviewer 4: > For `variant` type, the `lower_bounds` and `upper_bounds` store the minimum and maximum values for all shredded subcolumns within a file.

I'm not quite sure if we need to add the collection behavior of `lower_bound` and `upper_bound` under different conditions as mentioned in the [proposal](https://docs.google.com/document/d/1gAvt0x_ez89O8y-YqkCdMnTEykb-583YslYOgzf5sPg/edit?usp=sharing). What do you think?

Author: @XBaith What kind of collection behavior you are referring to? Can you clarify?

Reviewer 4: > Lower and upper bound statistics for subcolumns are collected for each data file based on the following conditions:
Uniform Value Types:
If all subcolumn values match the shredded type, lower/upper bounds are collected.
Example: For event.location.longitude, if all values are of the double type, the lower/upper bounds are written to the manifest file.
Mixed Value Types:
If the subcolumn contains multiple types (e.g., double and string), lower/upper bound statistics are not collected.
Example: For event.location.longitude, if the values include both double and string, lower/upper bounds are excluded.
Some subcolumn values are nulls or missing:
If some subcolumn values are null or missing in a file, but the available values match the shredded type, lower/upper bounds are still collected.
If all the subcolumn values are nulls, then lower/upper bounds are not collected. null_value_counts stat can be collected in later implementation to be used with value_counts to know they are all nulls.

Author: Make sense. Let me include that: basically we only collect bounds when all the subcolumn values are of the same shredded type or null.

> Subcolumn paths follow the JSON path format, such as `$.event.event_type` for standard keys or `$.event.["user.name"]` for keys containing dots.

Author: Make sense. That would reduce the variation. I also update to add the array representation to use always index 0 like  `$['tags'][0]`. Let me know your thoughts.

> For `variant` type, the `lower_bounds` and `upper_bounds` store the lower and upper bounds for the shredded fields within a file with the following considerations: 1) Bounds for array data are not collected; 2) The lower / upper bounds are collected only if all field data share the same shredded type or if the data is missing. These bounds are represented as a Variant object, where each field path serves as a key and the corresponding bound value as the value. The object is then serialized into binary format (see [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md)).

Reviewer 2: ```suggestion
For Variant, values in the `lower_bounds` and `upper_bounds` maps store the a serialized Variant object that contains lower and upper bounds for fields within the Variant. The object keys are normalized JSON path expressions that uniquely identify a Variant field. The object values are primitive Variant representations of the lower or upper bound for the field. Including bounds for any field is optional and the bounds must have the same Variant type.

Bounds for a field must be accurate for all non-null values of the field in a data file. Bounds for values within arrays must be accurate all values in the array. Bounds must not be written to describe values with mixed Variant types (other than null). For example, a "measurement" field that contains int64 and null values may have bounds, but a string value such as "n/a" or "0" in any record would cause the bounds to be skipped.

The Variant bounds objects are serialized by concatenating the [Variant encoding](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md) of the metadata (containing the normalized field paths) and the bounds object.
```

Author: `the bounds must have the same Variant type.` is not clear to me.  

We are expecting "the field values must have the same Variant type." right?

> Subcolumn paths follow the JSON path format to use normalized path, such as `$['location']['latitude']` or `$['user.name']`. If the shredded subcolumn is an array, represent it using the index 0 to indicate the array structure, such as `$['tags'][0]`.

Reviewer 2: There shouldn't be a need to include array paths. Bounds for array data are not tracked.

|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **`geometry`**               | A single point, encoded as a x:y:z:m concatenation of its 8-byte little-endian IEEE 754 coordinate values. x and y are mandatory. This becomes x:y if z and m are both unset, x:y:z if only m is unset, and x:y:NaN:m if only z is unset. |
| **`geography`**              | A single point, encoded as a x:y:z:m concatenation of its 8-byte little-endian IEEE 754 coordinate values. x and y are mandatory. This becomes x:y if z and m are both unset, x:y:z if only m is unset, and x:y:NaN:m if only z is unset. |

> | **`variant`**              | A `Variant` object, where each subcolumn path serves as a key and the corresponding bound value as the value. Subcolumn paths follow the JSON path format. |

Reviewer 2: `Variant` is not something that we can reference here because that is a class in the implementation. I think what you want to say is that the serialized value is a variant metadata (v1) concatenated with a variant object. The object's fields are field paths in the normalized JSON path format and the values are the upper or lower bound corresponding to the shredded type in the data file.

Author: Make sense. Here we want to describe how exactly the data is laid out

> | **`variant`**                | The serialized value consists of a variant metadata(v1) concatenated with a variant object. The object's fields are field paths in the normalized JSON path format and the values are the upper or lower bounds for the corresponding shredded type in the data file. |

Reviewer 2: ```suggestion
| **`variant`**                | A serialized Variant of encoded v1 metadata concatenated with an encoded variant object. Object keys are normalized JSON paths to identify fields; values are lower or upper bound values. |
```
